### PR TITLE
Add `MessageEntity.shift_entities` and `MessageEntity.concatenate`

### LINKED ISSUE #4372: [FEATURE] Convenience Method for `MessageEntity` Concatenation
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

The discussion with TG in https://github.com/tdlib/telegram-bot-api/issues/515 has made it clear that we should encourage users to work more with MessageEntities rather than with our `text_html/markdown` helpers. Since this is sometimes cumbersome, convenience functionality for working with ME is surely advisable and we've made a step into that direction with #4323 .

What I'd like to see in addition is convenience functionality for catenating text with entities as initially touched in https://github.com/tdlib/telegram-bot-api/issues/515#issuecomment-1885594214

### Describe the solution you'd like

new class methods

* `ME.shift_entities(entities: Sequence[ME], by: str | int) -> Sequence[ME]`: shifts the entities by the specified length or the utf-16 length of the string of `by` is a string. This would be useful for adding a prefix to a message with entities
* `ME.concatetane(first: tuple[str, Sequence[ME]], second: tuple[str, Sequence[ME]]) -> tuple[str, Sequence[ME]]`: Concatenates text and entities. Uses `ME.shift_entities` internally

### Describe alternatives you've considered

`ME.concatenate` could be skipped in a first step as it can be implemented using `shift_entities`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
closes #4372 

The implemented signature for `ME.concatenate` differs from the one proposed in #4372 to allow concanetating several tuples of text & entities at once.

- [x] Added ``.. versionadded:: NEXT.VERSION``, ``.. versionchanged:: NEXT.VERSION``, ``.. deprecated:: NEXT.VERSION`` or ``.. versionremoved:: NEXT.VERSION`` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [x] Created new or adapted existing unit tests
- [ ] Documented code changes according to the `CSI standard <https://standards.mousepawmedia.com/en/stable/csi.html>`__
- ~Added myself alphabetically to ``AUTHORS.rst`` (optional)~
- ~Added new classes & modules to the docs and all suitable ``__all__`` s~
- [x] Checked the `Stability Policy <https://docs.python-telegram-bot.org/stability_policy.html>`_ in case of deprecations or changes to documented behavior

### CODE REVIEW

File: telegram/_messageentity.py [hunk 1]
```diff
@@ -219,6 +219,147 @@ def adjust_message_entities_to_utf_16(
...
+        Returns:
+            Tuple[:obj:`str`, Sequence[:class:`telegram.MessageEntity`]]: The concatenated text
+            and its entities
+        """
+        output_text = ""
+        output_entities: List[MessageEntity] = []
+        for arg in args:
+            text, entities = arg[0], arg[1]
+
+            if len(arg) > 2 and arg[2] is True:
```
Comment (elpekenin) [inline telegram/_messageentity.py:355]: Given users are supposed to pass bools (opposed to any other value that could be `True`-y), do we really need to explicitly `== True`?

File: telegram/_messageentity.py [hunk 2]
```diff
@@ -219,6 +219,147 @@ def adjust_message_entities_to_utf_16(
...
+            with new_entity._unfrozen():
+                new_entity.offset += effective_shift
+            out.append(new_entity)
+        return out
+
+    @classmethod
+    def concatenate(
+        cls,
+        *args: Union[
+            Tuple[str, Sequence["MessageEntity"]], Tuple[str, Sequence["MessageEntity"], bool]
```
Comment (elpekenin) [inline telegram/_messageentity.py:291]: Maybe worth adding an alias for the `tuple[str, Sequence[ME]]` type

**Review by elpekenin — APPROVED** [review]
Just a couple of nitpicks, LGTM

File: telegram/_messageentity.py (hunk 2 above)
Comment (Bibo-Joshi) [reply telegram/_messageentity.py:291]: maybe not for that (it's used only twice), but for `Sequence[ME]` (which is used 7 times :))

**Review by harshil21 — COMMENTED** [review]
cool idea, I like it :)

File: telegram/_messageentity.py [hunk 3]
```diff
@@ -220,6 +220,143 @@ def adjust_message_entities_to_utf_16(
...
+        Returns:
+            Tuple[:obj:`str`, Sequence[:class:`telegram.MessageEntity`]]: The concatenated text
+            and its entities
+        """
+        output_text = ""
+        output_entities: List[MessageEntity] = []
+        for arg in args:
+            text, entities = arg[0], arg[1]
+
+            if len(arg) > 2 and arg[2]:
```
[the remark this answers is not in the document]
Comment (harshil21) [reply telegram/_messageentity.py]: oh I just saw elpekenin review the same point ooops

File: telegram/_messageentity.py [hunk 4]
```diff
@@ -220,6 +220,143 @@ def adjust_message_entities_to_utf_16(
...
+            The :paramref:`entities` are *not* modified in place. The function returns a sequence
+            of new objects.
+
+        .. versionadded:: NEXT.VERSION
+
+        Args:
+            by (Union[:obj:`str`, :obj:`int`]): Either the amount to shift the offset by or
+                a string whose length will be used as the amount to shift the offset by. In this
+                case, UTF-16 encoding will be used to calculate the length.
+            entities (Sequence[:class:`telegram.MessageEntity`]): Sequence of entities
```
[the remark this answers is not in the document]
Comment (harshil21) [reply telegram/_messageentity.py:271]: well yeah, but the description of the parameter just looked a bit too short

**Review by harshil21 — APPROVED** [review]

**Pull request merged**